src/Microsoft.FeatureManagement/ConfigurationFeatureDefinitionProvider.cs

```diff
@@ -113,16 +116,53 @@ public async IAsyncEnumerable<FeatureDefinition> GetAllFeatureDefinitionsAsync()
             // Iterate over all features registered in the system at initial invocation time
             foreach (IConfigurationSection featureSection in GetFeatureDefinitionSections())
             {
+                string featureName = GetFeatureFlagSectionName(featureSection);
+
+                if (string.IsNullOrEmpty(featureName))
+                {
+                    continue;
+                }
+
                 //
                 // Underlying IConfigurationSection data is dynamic so latest feature definitions are returned
-                yield return  _definitions.GetOrAdd(featureSection.Key, (_) => ReadFeatureDefinition(featureSection));
+                yield return  _definitions.GetOrAdd(featureName, (_) => ReadFeatureDefinition(featureSection));
+            }
+        }
+
+        private void EnsureInit()
+        {
+            if (!_configuration.GetChildren().Any())
+            {
+                Logger?.LogDebug($"Configuration is empty.");
+            }
+
+            bool hasFeatureManagementSection = _configuration.GetChildren()
+                .Any(section => string.Equals(section.Key, ConfigurationFields.FeatureManagementSectionName, StringComparison.OrdinalIgnoreCase));
+
+            if (!hasFeatureManagementSection && !RootConfigurationFallbackEnabled)
+            {
+                Logger?.LogDebug($"No configuration section named '{ConfigurationFields.FeatureManagementSectionName}' was found.");
+            }
+
+            if (_initialized == 0)
+            {
+                IConfiguration featureManagementConfigurationSection = hasFeatureManagementSection ?
+                    _configuration.GetSection(ConfigurationFields.FeatureManagementSectionName) :
+                    _configuration;
+
+                bool hasAzureAppConfigurationFeatureFlagSchema = HasAzureAppConfigurationFeatureFlagSchema(featureManagementConfigurationSection);
+
+                if (Interlocked.Exchange(ref _initialized, 1) != 1)
```

Comment:
Interlocked usage here can allow a simultaneously executing thread to bypass the critical section and return from the method without initializing. With the requirement to initialize here a lock cannot be avoided. Single usage of lock on initialization should be acceptable.